feat(flags): Implement a project specific token for feature flag definitions (aka local evaluation)

haacked opened the pull request:
## Problem

To use local evaluation and feature flag remote config (not to be confused with the other remote config), clients need to make a request using a *personal api key* to our `local_evaluation` or `remote_config` endpoint. The problem with this is that the personal api key is associated with a user and not a project (aka `team_id`). So if the user leaves the organization, the token no longer works.

## Changes

This implements a new team-level secret token specifically for retrieving feature flag definitions for local evaluation. Specifically, it adds two new properties to `team` (see #31876 for the PR that added the properties and migration).

1. `secret_api_token` - used for local evaluation and flags remote config.
2. `secret_api_token_backup` - also used for local evaluation and encrypted remote config.

### Why two columns?

Originally, we were only going to have one column to keep the implementation simple. But this removes a feature that otherwise exists with personal api keys, the ability to roll them.

We could do this with an array column or a new table, but that complicates the code. We want to keep this change simple.

We considered a more general approach, but now is not the time.

### UI Changes

New setting in Feature Flags Settings (I'm open to moving this to Team settings if people feel strongly). Note that we don't prepopulate this token. So this is what it'll look like the first time someone goes to get a token.

<img width="1457" alt="Screenshot 2025-05-06 at 12 24 37 PM" src="https://github.com/user-attachments/assets/0edd3529-5d03-4e6c-9a7b-5ce470a7d48a" />

The first time they generate a token, the dialog says this:

<img width="968" alt="Screenshot 2025-05-06 at 12 24 56 PM" src="https://github.com/user-attachments/assets/4d91b57c-ed61-4b5f-ad17-b827bc7c5f5e" />

Now they have a single "primary" key.

<img width="968" alt="Screenshot 2025-05-06 at 12 25 16 PM" src="https://github.com/user-attachments/assets/bf6aa0de-9ee4-4398-829b-cd7f11a8c253" />

If they need to rotate the key, they can click the rotate button. This shows a dialog that provides details on what's going to happen. I want to make it clear their existing code will continue to work.

<img width="1177" alt="Screenshot 2025-05-06 at 12 25 30 PM" src="https://github.com/user-attachments/assets/a3016f8f-588c-4bcf-b3bf-76daee029a80" />

When they rotate the key, we copy the old key into the "backup" key and generate a new primary key. And we note that this old key should be deleted after they deploy the new key.

<img width="974" alt="Screenshot 2025-05-06 at 12 25 38 PM" src="https://github.com/user-attachments/assets/c2a5bf55-c41f-48bf-920c-ca791ebef9a0" />

At any time, they can delete the backup key.

<img width="1051" alt="Screenshot 2025-05-06 at 12 25 46 PM" src="https://github.com/user-attachments/assets/00466d6b-4c69-47bc-9ce4-f07d896c9fcb" />

We also have activity log entries for these actions. You can see the sequence going from the bottom up.

![Screenshot 2025-05-06 at 4 27 01 PM](https://github.com/user-attachments/assets/e9ea1303-ef8d-434b-94a5-c214542556ac)


## Does this work well for both Cloud and self-hosted?

Yes.

## How did you test this code?

Unit tests and manually.

haacked committed:
Implement a ProjectSecretAPIKeyAuthentication

This is an authenticator for the ProjectSecretAPIKey

haacked committed:
Update local_evaluation to support project secret api token

And ensure that they continue to work for personal api keys.

haacked committed:
Add default value to mock

haacked committed:
Implement UI to rotate the secret tokens

haacked committed:
Correct copy pasta spelling error

haacked commented on frontend/src/scenes/feature-flags/FeatureFlagSettings.tsx:
```diff
@@ -39,10 +42,128 @@ export function FeatureFlagSettings({ inModal = false }: FeatureFlagSettingsProp
+                        }
+                        thing="Backup Feature Flags Secure API key"
+                    >
+                        {currentTeam.secret_api_token_backup}
```
We already checked earlier. This is in the branch where `currentTeam?.secret_api_token_backup` is not undefined.

haacked commented on frontend/src/scenes/teamActivityDescriber.tsx:
```diff
@@ -30,6 +30,24 @@ const teamActionsMapping: Record<
             description: [<>{prefix} the project API key</>],
         }
     },
+    secret_api_token: (change) => {
+        if (change === undefined || change.after === undefined) {
+            return null
+        }
+        const prefix = change.action === 'created' ? 'generated' : 'rotated'
+        return {
+            description: [<>{prefix} the Feature Flags secure API key</>],
+        }
+    },
```
Not needed.

haacked commented on posthog/api/test/test_authentication.py:
```diff
@@ -950,3 +955,83 @@ def test_after_timeout_modifications_require_reauthentication(self):
+
+    def test_authenticate_with_valid_secret_api_key_in_body(self):
+        # Simulate a request with a valid secret API key
+        wsgi_request = self.factory.post(
+            "/",
+            data=f'{{"secret_api_key": "{self.team.secret_api_token}"}}',
+            content_type="application/json",
+        )
```
This is a unit test. It's fine.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

haacked committed:
Add migration to ensure secret_api_token is unique

haacked commented on posthog/models/team/team.py:
```diff
@@ -165,6 +167,21 @@ def get_team_from_cache_or_token(self, token: Optional[str]) -> Optional["Team"]
         except Team.DoesNotExist:
             return None
 
+    def get_team_from_cache_or_secret_api_token(self, secret_api_token: Optional[str]) -> Optional["Team"]:
+        if not secret_api_token:
+            return None
+        try:
+            team = get_team_in_cache(secret_api_token)
+            if team:
+                return team
+
+            team = Team.objects.get(secret_api_token=secret_api_token)
+            set_team_in_cache(secret_api_token, team)
+            return team
+
+        except Team.DoesNotExist:
+            return None
```
Done.

haacked committed:
Update verifiedDomainsLogic.test.ts.snap

haacked committed:
Add secret_api_token to project.

haacked committed:
Refactor common check into a method.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (7)

haacked committed:
Remove unused snapshot

haacked committed:
Refactor actions that require current team into a separate list

haacked commented on posthog/api/team.py:
```diff
@@ -828,7 +852,11 @@ def event_ingestion_restrictions(self, request, **kwargs):
+            self.get_object()
+            if self.action in ["delete_secret_token_backup", "reset_token", "rotate_secret_token"]
+            else None
+        )
```
Agreed. Done.

haacked commented on posthog/auth.py:
```diff
@@ -175,6 +175,84 @@ def authenticate_header(cls, request) -> str:
+        self.team = team
+        self.current_team_id = team.id
+        self.is_authenticated = True
+        self.pk = 0
```
For what this is used for, it doesn't matter.

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (7)

haacked committed:
Implement the required authenticate_header method

Copied the implementation from `PersonalAPIKeyAuthentication`

haacked commented on posthog/auth.py:
```diff
@@ -175,6 +175,84 @@ def authenticate_header(cls, request) -> str:
         return cls.keyword
 
 
+class ProjectSecretAPIKeyAuthentication(authentication.BaseAuthentication):
```
Done.

haacked committed:
Add unit test and more assertions

haacked commented on posthog/api/test/test_team.py:
```diff
@@ -600,6 +600,242 @@ def test_reset_token_insufficient_priviledges(self):
+            response = self.client.patch(f"/api/environments/{self.team.id}/rotate_secret_token/")
+            self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+            # Make sure it's unchanged
+            self.assertEqual(self.team.secret_api_token, "phs_JVRb8fNi0XyIKGgUCyi29ZJUOXEr6NF2dKBy5Ws8XVeF11C")
```
Done.

haacked committed:
Add clarifying comment

haacked committed:
Remove the IS NOT NULL.

Django migrations --check complains about it. And it's not strictly necessary because `null` values are considered unique.

benjackwhite commented on posthog/api/test/test_authentication.py:
```diff
@@ -950,3 +955,83 @@ def test_after_timeout_modifications_require_reauthentication(self):
+
+    def test_authenticate_with_valid_secret_api_key_in_header(self):
+        # Simulate a request with a valid secret API key
+        wsgi_request = self.factory.get(
```
You can also do `CONFIG_AUTO_LOGIN` which clears the session so you can continue to use the normal client. Feels like it would be simpler and this is what we do elsewhere for testing api keys for example [here](https://github.com/PostHog/posthog/blob/master/posthog/api/test/test_personal_api_keys.py#L231)

haacked replied:
Correct me if I'm wrong, but I'm not sure that applies to these tests because I'm testing the authenticator directly. I need a request to pass into the `authenticate` method.

I do have other tests that test the authentication in the context of the full request pipeline.

benjackwhite reviewed (changes requested):
Minor changes. If you disagree with them and want to move forward then re-request and ill stamp otherwise i dont think they would be too hard and would make this much more future-friendly

zlwaterfield commented on posthog/api/project.py:
```diff
@@ -573,6 +579,32 @@ def reset_token(self, request: request.Request, id: str, **kwargs) -> response.R
+        methods=["PATCH"],
+        detail=True,
+        # Only ADMIN or higher users are allowed to access this project
+        permission_classes=[TeamMemberStrictManagementPermission],
```
This is looking at org membership levels so shouldn't be affected by the new access control except for the use of `team.effective_membership_level`, which I have this PR (https://github.com/PostHog/posthog/pull/31986) that I'll get shipped today adds access control support for the effective_membership_level. So TLDR; _should_ be fine.

haacked commented on posthog/permissions.py:
```diff
@@ -510,6 +527,11 @@ def has_permission(self, request, view) -> bool:
+        if isinstance(request.user, ProjectSecretAPIKeyUser):
+            # NOTE: If the user is a ProjectSecretAPIKeyUser then we've already checked that they have access to the team
+            # and that this request is for the local_evaluation endpoint.
+            return _is_request_for_project_secret_api_token_secured_endpoint(request)
```
Good call. I got here through a circuitous route, but my original assumptions no longer apply so a custom permission makes way more sense now.

haacked committed:
Add a custom permission for project secret api token

We still have to update the other permissions to ignore requests authenticated by our new authentication method
because permissions are additive. So to add this new custom permission, we're still calling all the other permissions.

I could factor out these two endpoints to their own ViewSet, but that's a lot of churn and I'm already porting local_evaluation to Rust.

haacked commented on posthog/permissions.py:
```diff
@@ -177,6 +178,10 @@ class TeamMemberAccessPermission(BasePermission):
     message = "You don't have access to the project."
 
     def has_permission(self, request, view) -> bool:
+        if isinstance(request.successful_authenticator, ProjectSecretAPIKeyAuthentication):
+            # Ignore the team check for project secret API keys. It's handled by the ProjectSecretAPITokenPermission
+            return True
```
Note: I still need to exclude `ProjectSecretAPIKeyAuthentication` from here because permissions are additive and this is still going to get called because this permission is registered in the `FeatureFlagViewSet`.

I considered moving `local_evaluation` and `remote_config` to a new `ViewSet`, but that's a lot more churn than I want to take on right now considering I'm also concurrently rewriting `local_evaluation` in Rust.

So hopefully this code here is temporary once we can move `local_evaluation`.

benjackwhite replied:
ah yeah you're right... We should just hurry up and move this endpoint to Rust and do it there 😆

haacked replied:
I'm working on it!
